### PR TITLE
Moving the dotprompt package into a plugin

### PR DESCRIPTION
**Pull request by maxl0rd**
- Dotprompt should now be configured as a plugin but has fallback behavior to look in a default prompts folder so no existing code should break.
- Async init used to load the entire prompts folder even if prompts are not explicitly referenced in code.

### CODE REVIEW

File: js/plugins/dotprompt/src/registry.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import { Dotprompt } from './prompt.js';
+
+export async function lookupPrompt(
+  dir: string,
+  name: string,
+  variant?: string
+): Promise<Dotprompt> {
+  // Expect to find the prompt in the registry
+  const registryPrompt = (await lookupAction(
+    `/prompt/${name}${variant ? `.${variant}` : ''}`
```
Comment (pavelgj) [inline js/plugins/dotprompt/src/registry.ts:31]: why not register .prompts under `/prompt/dotprompt/${name}${variant ? `.${variant}` : ''}`? This way `lookupAction` will handle plugin initialization for you.

Comment (maxl0rd) [reply js/plugins/dotprompt/src/registry.ts:31]: Ok, I wasn't aware of that, but I see how it resolves some issues. I'll try that.

File: js/plugins/dotprompt/src/registry.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  const expectedFileName = `${name}${variant ? `.${variant}` : ''}.prompt`;
+  const promptFolder = resolve(dir);
+  const promptExists = existsSync(join(promptFolder, expectedFileName));
+  if (promptExists) {
+    return loadPrompt(promptFolder, expectedFileName);
+  } else {
+    throw new GenkitError({
+      source: 'dotprompt',
+      status: 'NOT_FOUND',
+      message: `Could not find '${name}.prompt' in the prompts folder.`,
```
Comment (mbleigh) [inline js/plugins/dotprompt/src/registry.ts:56]: To cover variants don't you want

```suggestion
      message: `Could not find '${expectedFileName}' in the prompts folder.`,
```

File: js/plugins/dotprompt/src/index.ts [hunk 3]
```diff
@@ -14,14 +14,53 @@
...
 import { basename } from 'path';
-
 import { defineDotprompt, Dotprompt } from './prompt.js';
-import { lookupPrompt } from './registry.js';
+import { loadPromptFolder, lookupPrompt } from './registry.js';
 
 export { defineDotprompt, Dotprompt };
 
+export interface DotpromptPluginOptions {
+  promptDir: string;
```
Comment (mbleigh) [inline js/plugins/dotprompt/src/index.ts:27]:
```suggestion
  dir: string;
```

It's the Dotprompt plugin, so the directory is naturally for prompts. Also this should have a default value of `./prompts`.

File: js/plugins/dotprompt/src/index.ts [hunk 4]
```diff
@@ -14,14 +14,53 @@
...
+}
+
+let pluginOptions: DotpromptPluginOptions;
```
Comment (mbleigh) [inline js/plugins/dotprompt/src/index.ts:30]: This is weird global state stuff. Will this go away once you fix async init?

File: js/plugins/dotprompt/src/registry.ts (hunk 2 above)
Comment (maxl0rd) [reply js/plugins/dotprompt/src/registry.ts]: Done

File: js/plugins/dotprompt/src/index.ts (hunk 3 above)
Comment (maxl0rd) [reply js/plugins/dotprompt/src/index.ts]: Done

File: js/plugins/dotprompt/src/index.ts (hunk 4 above)
Comment (maxl0rd) [reply js/plugins/dotprompt/src/index.ts:34]: Yes, we don't need this any more!

File: js/plugins/dotprompt/src/registry.ts (hunk 1 above)
Comment (maxl0rd) [reply js/plugins/dotprompt/src/registry.ts]: This is working now!

File: js/plugins/dotprompt/src/index.ts (hunk 4 above)
Comment (mbleigh) [reply js/plugins/dotprompt/src/index.ts:34]: Can you remove it then?

Comment (maxl0rd) [reply js/plugins/dotprompt/src/index.ts]: Yes, it's removed. I just forgot to push it.

File: js/plugins/dotprompt/src/registry.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    });
+  }
+}
+
+export async function loadPromptFolder(
+  dir: string = './prompts'
+): Promise<void> {
+  const promptsPath = resolve(dir);
+  return new Promise<void>((resolve, reject) => {
+    readdir(
```
Comment (pavelgj) [inline js/plugins/dotprompt/src/registry.ts:74]: there's a `readdirSync` variant... that you can use to avoid all this boilerplate with Promises.

**Review by pavelgj — APPROVED** [review]

**Comment by MichaelDoyle** [comment]
Any samples that need to get updated based on this new plugin?

File: js/plugins/dotprompt/src/registry.ts (hunk 5 above)
Comment (maxl0rd) [reply js/plugins/dotprompt/src/registry.ts:74]: I looked at that, but it was slightly harder to resolve the initialization at the right place.

**Pull request merged**